### ISSUE TITLE
Being able to read PKCS12 'friendly name'

### ISSUE THREAD

**Issue by felixfontein** [pyca/cryptography#5760]
Right now, the interface for loading PKCS12 files (https://cryptography.io/en/latest/hazmat/primitives/asymmetric/serialization.html?highlight=pkcs12#cryptography.hazmat.primitives.serialization.pkcs12.load_key_and_certificates) returns a tuple `(private_key, certificate, additional_certificates)`. This tuple does not contain the 'friendly name', and assuming you don't want to have a different tuple length depending on some function argument, this cannot be properly extended.

Goal: make it possible to load the PKCS12 friendly name.

Idea: how about creating a small immutable object for PKCS12 content (i.e. it contains private key, certificate, list of additional certificates, friendly name - and maybe more in the future?) that can be serialized and deserialized by a new set of functions?

Does this sounds good? Does anyone have a better idea? I'm hesistant of suggestion a completely new interface, now that we "just" got PKCS12 support a few versions ago. But I can't think of a much better way to be able to return more data (friendly name now) while keeping backwards compatibility.

**Comment by felixfontein** [comment]
@reaperhulk I guess this would be what you originally proposed in https://github.com/pyca/cryptography/issues/2860#issuecomment-433680572 and what you implemented in #4553 until https://github.com/pyca/cryptography/compare/b2b793f01ba187ad887b1868898848cb43a43053..b6e8f170b0c3a4764017d4bf6f7f060c9ba08003 / https://github.com/pyca/cryptography/pull/4612#issuecomment-441614062.

**Comment by reaperhulk** [comment]
The existence of the current API doesn't preclude adding a more fully-featured API for people who need more complexity. In a parsed object how do you envision the friendly name being accessed though? e.g. if I have a `PKCS12` object that contains two certificates, how do I get the associated value? Can friendly names be associated with a private key?

**Comment by felixfontein** [comment]
That's a very good question. It depends a lot on how fully-featured the API should be. For me, it would be enough if the API allows to extract the same information that were used during serialization with cryptography, i.e. what OpenSSL's `PKCS12_create()` does: the friendly name of the parsed PKCS12 would be the friendly name associated with the serialized key+certificate pair (or more precisely, of the certificate that's part of that pair). (And if there is no key, there's no friendly name.)

That would be essentially the same what PyOpenSSL's `PKCS12` offers.

(I guess everything more complicated would make parsing / serializing more complicated, since `PKCS12_parse()` and `PKCS12_create()` can no longer be used.)

**Comment by reaperhulk** [comment]
I think this is another situation where perfect is the enemy of good. We will paralyze ourselves forever if we attempt to design an API that handles all the complexity of PKCS12 (because it is a fractal of insanity). What's the minimum API we can do here? Is it safe to do `load_key_and_certificates_with_friendly_name` and return a four element tuple? Can there be more than one friendly name?

**Comment by Ajpantuso** [comment]
I don't think it's a matter of changing the interface.

`PKCS12_parse()` is already mapping the `friendlyName` onto each certificate (See [here](https://www.openssl.org/docs/man1.1.1/man3/PKCS12_parse.html))

> NOTES
> The parameters pkey and cert cannot be NULL. ca can be <NULL> in which case additional certificates will be discarded. *ca can 
> 
> also be a valid STACK in which case additional certificates are appended to *ca. If *ca is NULL a new STACK will be allocated.
> 
> The friendlyName and localKeyID attributes (if present) on each certificate will be stored in the alias and keyid attributes of the X509 structure.

However `alias` is not a part of the certificate, but the greater x509 structure within the `trust settings` which are not accessible by any convenient means and are excluded in the default PEM representation (See [here](https://www.openssl.org/docs/man1.0.2/man3/PEM_write_bio_X509.html))

> The X509 functions process an X509 certificate using an X509 structure. They will also process a trusted X509 certificate but any trust settings are discarded.

There looks to be an alternative function which can handle `trust settings` however. (See [here](https://www.openssl.org/docs/man1.0.2/man3/PEM_write_bio_X509_AUX.html))

> The X509_AUX functions process a trusted X509 certificate using an X509 structure.

But I think making the `alias` accessible from each `certificate` object is the correct approach. (Or maybe that should be reserved for a new `TrustedCertificate` class assuming OpenSSL's trusted certificates can be consistently identified)

**Comment by Ajpantuso** [comment]
I did a little more digging and there are in fact specific functions which can retrieve the trusted settings. (See [here](https://github.com/openssl/openssl/blob/a227ff336fc934b4a52c4659ae423dfa10efd5f3/crypto/x509/x_x509a.c#L85-L101))

alias (friendlyName) -> X509_alias_get0
keyid (localKeyID) -> X509_keyid_get0

In particular X509_alias_get0 is already included in `cryptography` so as a temporary hack @felixfontein it would be possible to do the following:

```python
from cryptography.x509 import load_pem_x509_certificate
from cryptography.hazmat.primitives.serialization.pkcs12 import load_key_and_certificates
from cryptography.hazmat.primitives.serialization import Encoding

with open('./any_pkcs12_file', 'rb') as p12:
    pkcs12 = load_key_and_certificates(
        p12.read(),
        password='password'.encode('UTF-8'),
    )

_, cert, _ = pkcs12

print(
    cert._backend._ffi.string(
        cert._backend._lib.X509_alias_get0(
            cert._x509, cert._backend._ffi.new("int *")
        )
    )
)
```

**Comment by felixfontein** [comment]
This workaround unfortunately stopped working (the most obvious reason why it stopped is that `cert._backend` no longer exists, but replacing it by a new instance of `default_backend()` doesn't work either - it looks like that information has been stripped from `cert._x509`).

**Comment by reaperhulk** [comment]
Yeah we need a public API here now because we no longer use OpenSSL for our x509.Certificate structures.

**Comment by felixfontein** [comment]
I've created a PR #6348 which provides a `load_key_and_certificates_with_name` function for this.
